src/v2/api/services/RoleService.ts

```diff
@@ -50,30 +57,108 @@ export class RoleService {
   }
 
-  update (id: string, body: UpdateRoleDTO) {
+  async update (id: string, body: UpdateRoleDTO) {
     return this.roleRepository.updateById(id, body);
   }
 
   get (roleId: string) {
     return this.roleRepository.findById(roleId);
   }
 
-  getAll () {
-    return this.roleRepository.findMany();
+  async getAll (query: QueryAllRolesDTO) {
+    const search = DatabaseUtils.getSearch(query, 'displayName');
+    const sort = DatabaseUtils.getSort(query, 'permission');
+
+    const data: Prisma.RoleFindManyArgs = {
+      where: {
+        ...search,
+        name: query.name,
+        groupRole: query.groupId ? {
+          groupId: query.groupId,
+        } : undefined,
+      },
+      ...sort,
+    };
+    const roles = await DatabaseUtils.paginate<DbRole>(this.grantRepository, query, data);
+
+    return {
+      data: roles.data,
+      pagination: roles.pagination,
+    };
+  }
+
+  async getGrants (roleId: string, query: QueryAllGrantsDTO) {
+    const search = DatabaseUtils.getSearch(query, 'permission');
+    const sort = DatabaseUtils.getSort(query, 'permission');
+
+    const data: Prisma.GrantFindManyArgs = {
+      where: {
+        ...search,
+        set: query.set,
+        roleId,
+      },
+      ...sort,
+    };
+    const grants = await DatabaseUtils.paginate<Grant>(this.grantRepository, query, data);
+
+    return {
+      data: grants.data,
+      pagination: grants.pagination,
+    };
+  }
+
+  async createGrants (roleId: string, grants: CreateGrantDTO[]) {
+    const createGrants = grants.map((g) => ({ roleId, ...g }));
+    for (const grant of createGrants) {
+      await this.updateHigherGrants(roleId, grant.weight);
+    }
+
+    return this.grantRepository.createMany(createGrants);
+  }
+
+  async createGrant (roleId: string, body: CreateGrantDTO) {
+    const grant = { ...body, roleId };
+    await this.updateHigherGrants(roleId, body.weight);
+
+    return this.grantRepository.create(grant);
   }
 
-  async getGrants (roleId: string) {
-    const role = await this.roleRepository.findById(roleId);
-    return role.grants;
+  async updateGrant (roleId: string, grantId: string, body: UpdateGrantDTO) {
+    await this.updateHigherGrants(roleId, body.weight);
+
+    return this.grantRepository.updateById(grantId, body);
+  }
+
+  private async updateHigherGrants (roleId: string, weight: number) {
+    const higherGrants = await this.grantRepository.findMany({
+      where: {
+        roleId,
+        weight: {
+          gte: weight,
+        },
+      },
+      orderBy: {
+        weight: 'asc',
+      },
+    });
+
+    const data: Prisma.GrantUpdateManyMutationInput[] = [];
+
+    for (let i = 0; i < higherGrants.length; i++) {
+      const grant = higherGrants[i];
+      grant.weight += 1;
+      data.push(grant);
+
+      if (i === higherGrants.length - 1 || grant.weight !== higherGrants[i + 1].weight) {
+        break;
+      }
+    }
+
+    this.grantRepository.updateMany(data);
```

Comment:
The function doesn't work correctly because you increase the weight of the first grant, even if it is > 1 than that of the initial one. Pls take a look
```js
let prev = weight;
for (let i = 0; i < higherGrants.length; i++) {
  const grant = higherGrants[i];
  if (prev !== grant.weight + 1) break;
  grant.weight++;
  data.push(grant);
  prev = grant.weight;
}
```